### PR TITLE
fix(native-rd): enforce Evolu Result handling at redesign mutation boundaries (#503)

### LINKED ISSUE #503: [Foundation] Enforce Evolu Result handling at redesign mutation boundaries
Part of **Epic #384 — Full Ride redesign**. Persistence prerequisite for #444, #446, #449, and #469.

## Problem

Several mutation callers catch thrown exceptions but ignore Evolu `{ ok: false }` Results. The UI can clear a draft, close confirmation, navigate away, or report bake completion even though persistence failed. Critical examples include badge creation/completion, goal/step edits, step deletion, badge deletion, and preference persistence.

Joe's product decision: the redesign retains full functionality and failure recovery; simplification must not create silent data loss.

## Scope

- Audit every mutation boundary exercised by the redesigned flows.
- Handle both thrown errors and `ok:false` Results.
- Introduce a small shared result-unwrapping helper only if it makes call sites simpler and consistently typed.
- Do not clear drafts, close destructive confirmation, advance stages, or navigate on rejected writes.
- Preserve user input and expose useful, non-cute error/retry feedback.
- Ensure multi-write flows define partial-failure behavior.

## Acceptance

- [ ] `useCreateBadge` cannot reach `done` unless badge persistence and goal completion succeed.
- [ ] New/Edit Goal retain drafts and remain on-screen after rejected writes.
- [ ] Delete flows do not navigate away after rejected deletion.
- [ ] Theme/density changes report or recover from persistence rejection rather than appearing permanently saved.
- [ ] Container/hook tests cover both thrown and `ok:false` paths.
- [ ] Errors are logged/reported with safe context and users receive actionable feedback.

**Estimate:** 300–550 changed LOC including tests.

### PR DESCRIPTION
**Pull request by joeczar**
## Summary

Enforce Evolu `Result` handling at every mutation boundary in the redesigned Full Ride flows. Evolu reports write failures two ways — a thrown exception **and** a returned `{ ok: false }` — and call sites were only handling one (or neither), so rejected writes silently fired success-only side effects (reaching a "done" bake state, clearing a draft, closing a delete-confirmation modal and navigating away, or leaving a preference looking "saved"). A shared `runEvoluMutation` helper normalizes both failure modes; every touched caller now gates its success-only side effect on the result.

## Changes

- **`src/utils/evoluMutation.ts` (new):** `runEvoluMutation(mutate, onFailure): boolean` — collapses thrown + `{ ok: false }` into one `onFailure(error)` and returns whether the write succeeded (D1).
- **`useCreateBadge`:** gate `setStatus("done")` on `createBadge`/`updateBadge`/`completeGoal` all returning `ok: true`; a rejected persist lands on `status: "error"` and never completes the goal.
- **`EditModeScreen`:** route every goal/step mutation through the helper; `handleDeleteGoal` now dismisses the modal + navigates only on success. (`reorderSteps`/`reorderSubSteps` already throw on failure and keep their existing `try/catch` — see plan Discovery Log.)
- **Delete flows:** `GoalsScreen.confirmDelete` (previously no error handling at all) and `BadgeDetailScreen.handleConfirmDelete` (dead catch; closed+navigated unconditionally) now keep the confirmation open and surface an error on a rejected soft-delete.
- **Preferences:** `useDensity.setDensity` / `useThemePersistence.setTheme` return a success boolean (they run above `<ToastProvider>`, D2); `SettingsScreen` and `ThemeSwitcher` show a toast on failure so nothing looks "saved" while nothing persisted.
- Added `{ area: "settings.theme" }` to the `ReportContext` union (D5); new `en` copy for goal-delete/density/theme failures (+ regenerated `pseudo/`; `de/` left to the i18n-sync bot).

## Dependencies

- [x] None

## Test Plan

- [x] Tests pass locally (`bun run test`) — 9976/9977; the one failure is a teardown-timeout flake in `EditGoalView` (untouched by this PR), which passes 110/110 in isolation
- [x] Lint passes (`bun run lint`) — 0 errors
- [x] Type check passes (`bun run type-check`)
- [x] Format check passes (`prettier --write` on all touched files)
- [ ] For visible UI changes: a11y audit — no new UI surfaces; only toast/alert feedback on existing controls via existing primitives

Every touched container/hook has a thrown-error test **and** a `{ ok: false }` test. Reviewer focus: the delete-flow reorderings (modal must stay open + no navigation on rejected delete) and the `useCreateBadge` partial-state edge case (badge row written but `completeGoal` fails → `status: "error"`, goal stays active).

## Related Issues

Closes #503

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: apps/native-rd/src/utils/evoluMutation.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * Preferred over Evolu's own `getOrThrow`, whose thrown message
+ * (`"getOrThrow"`) is non-descriptive and whose JSDoc explicitly discourages it
+ * for application error handling. Generalizes the `if (!result.ok)` pattern
+ * already used in `EditModeScreen`.
+ */
+export function runEvoluMutation<T, E>(
+  mutate: () => Result<T, E>,
+  onFailure: (error: E | unknown) => void,
+): boolean {
```
[the remark this answers is not in the document]
Comment (joeczar) [reply apps/native-rd/src/utils/evoluMutation.ts]: Fixed in d8bbee37 — `onFailure` is now `(error: unknown) => void` and the misleading `E` generic is gone; the helper takes `Result<unknown, unknown>`.

File: apps/native-rd/src/hooks/useTheme.ts [hunk 2]
```diff
@@ -48,7 +48,10 @@ interface ThemeContextValue {
   theme: ComposedTheme;
   isDark: boolean;
   variant: Variant;
-  setTheme: (name: ThemeName) => void;
+  // Returns false only when persisting the theme to Evolu failed, so the
+  // calling UI can surface a toast (#503). The outer-provider stub throws, and
+  // the pre-Evolu Unistyles-only provider returns void — both remain assignable.
+  setTheme: (name: ThemeName) => boolean;
```
[the remark this answers is not in the document]
Comment (joeczar) [reply apps/native-rd/src/hooks/useTheme.ts:55]: Fixed in d8bbee37 — the comment now states the real assignability reason (the outer stub throws, so its `never` return is assignable) and no longer claims `void` implementations are compatible.

File: apps/native-rd/src/hooks/useCreateBadge.ts [hunk 3]
```diff
@@ -346,21 +346,36 @@ export function useCreateBadge(
...
+              goalId,
+              credential: credentialJsonOut,
+              imageUri,
+              ...(designRef.current ? { design: designRef.current } : {}),
+            });
+        if (!badgeResult.ok) {
+          throw new Error(
+            `Failed to ${existingBadge ? "update" : "create"} badge record: ${JSON.stringify(badgeResult.error)}`,
+          );
+        }
```
[the remark this answers is not in the document]
Comment (joeczar) [reply apps/native-rd/src/hooks/useCreateBadge.ts:372]: Fixed in d8bbee37 — message is now the stable `Failed to update/create badge record`; the Evolu error rides on `{ cause }` for logs/Sentry only.

File: apps/native-rd/src/hooks/useCreateBadge.ts [hunk 4]
```diff
@@ -346,21 +346,36 @@ export function useCreateBadge(
...
+        // completeGoal may also fail after the badge row was written. We accept
+        // that partial state (badge exists, goal still active) rather than roll
+        // back — the goal stays active and the user can re-attempt — but we must
+        // NOT report "done" when the completion write itself failed.
+        const completeResult = completeGoal(goalId, goalEvidenceForGating);
+        if (!completeResult.ok) {
+          throw new Error(
+            `Failed to complete goal: ${JSON.stringify(completeResult.error)}`,
+          );
         }
```
[the remark this answers is not in the document]
Comment (joeczar) [reply apps/native-rd/src/hooks/useCreateBadge.ts:384]: Fixed in d8bbee37 — message is now the stable `Failed to complete goal` with the Evolu error on `{ cause }`.

File: apps/native-rd/src/screens/GoalsScreen/__tests__/GoalsScreen.test.tsx [hunk 5]
```diff
@@ -444,6 +453,54 @@ describe("GoalsScreen", () => {
...
+          }),
+      },
+      {
+        label: "thrown error",
+        arm: () =>
+          deleteGoal.mockImplementation(() => {
+            throw new Error("Failed to delete goal. Please try again.");
+          }),
+      },
+    ])("keeps the modal open and reports when delete fails (%s)", ({ arm }) => {
```
[the remark this answers is not in the document]
Comment (joeczar) [reply apps/native-rd/src/screens/GoalsScreen/__tests__/GoalsScreen.test.tsx]: Fixed in d8bbee37 — table rows carry a `label` key and the test name uses `($label)`. Same change applied to the EditModeScreen table.

File: apps/native-rd/src/components/ThemeSwitcher/ThemeSwitcher.tsx [hunk 6]
```diff
@@ -154,7 +156,16 @@ export function ThemeSwitcher() {
           return (
             <Pressable
               key={option.id}
-              onPress={() => setTheme(option.id)}
+              onPress={() => {
+                // setTheme returns false only when the persist failed; surface
+                // a toast so the option doesn't look selected while nothing
+                // persisted.
```
[the remark this answers is not in the document]
Comment (joeczar) [reply apps/native-rd/src/components/ThemeSwitcher/ThemeSwitcher.tsx]: Fixed in d8bbee37 — the comment now says the option still shows as selected (the theme did change in-session) and the toast tells the user the choice won't survive a restart.
